### ISSUE TITLE
liveness check causes double Auth

### ISSUE THREAD

**Issue by Mat-Hartje** [phpredis/phpredis#2536]
### Expected behaviour
When using auth and `redis.pconnect.echo_check_liveness`, should only need to AUTH once.

I'd expect phpredis to issue something like:
* AUTH (redacted)
* ECHO phpredis:66b54a22d8b02:79cc530d
* GET "mykey"

### Actual behaviour
Using monitor, I see:
* AUTH (redacted)
* ECHO phpredis:66b54a22d8b02:79cc530d
* AUTH (redacted)
* GET "mykey"

### I'm seeing this behaviour on
- OS: Windows and AL2023
- Redis: Elasticache
- PHP: 8.2.18
- phpredis: 6.1.0RC1

### Steps to reproduce, backtrace or example script
ini_set('redis.pconnect.echo_check_liveness',1)
$parameters = [ null, ['***.use1.cache.amazonaws.com:6379'], 1.5, 1.5, true,'QL645ZECfZXgA6eoQ0uy', [] ];
$redis = new RedisCluster(...$parameters); 
$redis->get('asdf');

Observe the commands that are issued.

### I've checked
- [x] There is no similar issue from other users
- [x] Issue isn't fixed in `develop` branch

**Comment by Mat-Hartje** [comment]
It looks like [this line](https://github.com/phpredis/phpredis/blob/develop/library.c#L3035) is intended to mark the connection as authenticated, but a [successful authentication results](https://github.com/phpredis/phpredis/blob/develop/library.c#L3031) in a short-circuit return before it can ever be reached.

**Comment by michael-grunder** [comment]
Thanks, I'll take a look tonight.

**Comment by Mat-Hartje** [comment]
As a slight aside, I'd love to know if you have any guidance around when to enable/disable `echo_check_liveness`?  It's not really mentioned in the docs. 

I started digging in here because of how chatty our app was being with redis. Turning the check off is pretty helpful, but from what I can glean from other Issues/PRs, it seems like the kind of setting that is only gonna be relevant under high load situations. (Not sure how much faith I can put into the fact that disabling the check seems fine in our dev environment)

Either way, thanks for looking into this.

**Comment by michael-grunder** [comment]
I do not see "double auth" but this feature can be improved.

Clearly we don't need to send `AUTH` if we get a response to our `ECHO`. I think the change should be small enough to get into the next release candidate.

I can put something together this weekend.

---

> As a slight aside, I'd love to know if you have any guidance around when to enable/disable echo_check_liveness? It's not really mentioned in the docs

It was a mitigation we added for users with persistent connections and very high traffic. 

Users were seeing "out of sync" data, as if there were still unprocessed replies in the receive buffer on the socket when the persistent link was retrieved from the pool. This caused all sorts of problems as you can imagine. Unfortunately I was never able to cause it locally, so I implemented the liveness check as a solution.

When enabled, we send the `ECHO` with a unique identifier after receiving the persistent link. As long as we get the echo response we know the connection is not out of sync.

Lots of people disable the setting so whether you should really just depends on whether or not it you end up seeing the above problem. If not, you can leave it off.

**Comment by Mat-Hartje** [comment]
Thanks @michael-grunder!

On the topic of `AUTH` and persistent connections, I'm also seeing an `AUTH` at the start of every request even when a persisted connection is getting re-used.  Is that expected?  I would have expected to only need to `AUTH` once during the lifetime of a connection.  (though maybe this is more of a quirk of laravel+phpredis)

**Comment by michael-grunder** [comment]
> AUTH at the start of every request even when a persisted connection is getting re-used

By "every request" do you mean every HTTP request? That's what I'm going to fix this weekend. What's happening is that we pull the connection from the persistent connection pool, send the `ECHO`, get a correct response and then don't infer from that we're authenticated. This means the first real command also triggers `AUTH`.

**Comment by yatsukhnenko** [comment]
@michael-grunder actually we send auth + ping and determine connection state on ping response is a good idea 👍

**Comment by michael-grunder** [comment]
There are a few edge cases we need to reason through to infer the authentication state.

The precise sequence happening now depends on a few factors.  How `pconnect` is called, and whether or not there is also an explicit call to `auth`.

Phpredis hashes persistent links by `host:port` or `/path/to/redis.sock` unless you use a non-default `redis.pconnect.pool_pattern`. If I had to guess, few people likely do this.

Also, whether or not `AUTH` is delivered on the liveness check will depend on how `pconnect` is called.

For example, if you do this:

```php
$redis = new Redis;
$redis->pconnect('localhost', 6379);
$redis->auth('secret');
$redis->ping();
```

You will see a sequence like this in your monitor output

```
1723316111.457078 [0 127.0.0.1:36888] "ECHO" "phpredis:66b7b78f6f913:0000001c"
1723316111.457249 [0 127.0.0.1:36888] "AUTH" "(redacted)"
1723316111.457374 [0 127.0.0.1:36888] "PING"
```

1. `ECHO` when we pull from the pool, 2. explicit `AUTH` call, 3. `PING`.

However, if you do something like this:

```php
$redis = new Redis;
$redis->pconnect('localhost', 6379, context: ['auth' => 'secret']);
$redis->ping();
```

You will see this sequence:

```
1723316261.959640 [0 127.0.0.1:36888] "AUTH" "(redacted)"
1723316261.959656 [0 127.0.0.1:36888] "ECHO" "phpredis:66b7b825ea449:0000001d"
1723316261.959785 [0 127.0.0.1:36888] "PING"
```

1. `AUTH` and `ECHO` when pulling from the pool, 2. `PING`.

You get the "double auth" if you do both:

```php
$redis = new Redis;
$redis->pconnect('localhost', 6379, context: ['auth' => 'secret']);
$redis->auth('secret');
$redis->ping();
```

```
1723316765.262947 [0 127.0.0.1:36888] "AUTH" "(redacted)"
1723316765.262960 [0 127.0.0.1:36888] "ECHO" "phpredis:66b7ba1d402dc:00000026"
1723316765.263079 [0 127.0.0.1:36888] "AUTH" "(redacted)"
1723316765.263178 [0 127.0.0.1:36888] "PING"
```

1. `AUTH` and `ECHO` pulling from the pool, 2. Explicit `AUTH`, 3. `PING`.

This is inefficient but the solution isn't straightforward, because what about this:

```php
$redis = new Redis;
$redis->pconnect('localhost', 6379, context: ['auth' => ['user1', 'password1']]);
$redis->get('user1-key');
$redis->auth('user2', 'password2');
$redis->get('user2-key');
```

And you could do this in any combination of operations. I don't think it's safe to change this in the current release.

We should fix it, but we need to think it through and implement a robust solution. We may want to change the default `redis.pconnect.pool_pattern` to include this auth information as a first step. This lets us be a bit more clever about when we need to deliver an explicit AUTH call.

**Comment by yatsukhnenko** [comment]
First of all IMO when user explicitly call some commands like `AUTH` we should always send them.
The place for optimization is connection pool and aliveness check where we can try to avoid sending `AUTH` before `ECHO`. Success response indicates that connection is fine and authenticated. In error case we can try to re-send `AUTH`+`ECHO` or do some other logic we want.
So for code
```php
$redis = new Redis;
$redis->pconnect('localhost', 6379, context: ['auth' => 'secret']);
$redis->auth('secret');
$redis->ping();
```
for the common case sequence should be
```
1723316765.262960 [0 127.0.0.1:36888] "ECHO" "phpredis:66b7ba1d402dc:00000026"
1723316765.263079 [0 127.0.0.1:36888] "AUTH" "(redacted)"
1723316765.263178 [0 127.0.0.1:36888] "PING"
```
For the worst case it might be two more commands (or not if we will use connection as is and expect that user control this)

> We may want to change the default redis.pconnect.pool_pattern to include this auth information as a first step

Nice, another possible solution is to wrap `php_stream` into some struct and keep additional metadata like auth state, db number, etc

> I don't think it's safe to change this in the current release.

Absolutely agree

**Comment by michael-grunder** [comment]
> Success response indicates that connection is fine and authenticated. In error case we can try to re-send AUTH+ECHO or do some other logic we want.

Right, that's what I'm thinking. We still need to send both auth and echo if the user passes auth info into the `pconnect` call though, because they could be sending a different username/password to the one the connection is logged in as.

So if the user omits the auth info and we get a good `ECHO` response we know we're authenticated.

If we extend the connection pool to store both the `php_stream` and some additional metadata we could omit the auth even if it was specifically passed to pconnect as we could tell that we are already authed to that same user.
